### PR TITLE
Feature/type traits

### PR DESCRIPTION
**Pull request by maddyscientist**
This pull request extends the support of `type_traits`, by adding initial support for the following C++11 features (except where noted they are C++14)
* `enable_if_t` (C++14)
* `is_integral`
* `is_signed`
* `is_unsigned`
* `is_array`
* `is_function`
* `remove_reference`
* `remove_reference_t`
* `remove_extent`
* `remove_extent_t` (C++14)
* `remove_const`
* `remove_const_t` (C++14)
* `remove_volatile`
* `remove_volatile_t` (C++14)
* `remove_cv`
* `remove_cv_t` (C++14)
* `conditional`
* `conditional_t` (C++14)
* `add_pointer`
* `add_pointer_t` (C++14)
* `decay`
* `decay_t` (C++14)

Moreover, the pull request also:
* Adds a mutex skeleton header for aiding cub compilation
* Adds proper error handling to `abi::__cxa_demangle` function call
* Adds CUDA arch detection during initial iterative compilation to ensure that correct internal intrinsics are included

### CODE REVIEW

**Comment by maddyscientist** [comment]
Noting that this pull request should be done after #14 is merged, since it is based from that branch.

**Comment by benbarsdell** [comment]
The diff here has not auto-updated since #14 was merged. Apparently it will if you change the destination to another branch and then change it back.

Alternatively you could rebase it.

**Comment by maddyscientist** [comment]
@benbarsdell done

File: jitify.hpp [hunk 1]
```diff
@@ -680,9 +680,18 @@ inline std::string demangle(const char* mangled_name) {
 inline std::string demangle(const char* mangled_name) {
   size_t bufsize = 1024;
   char* buf = (char*)malloc(bufsize);
```
Comment (benbarsdell) [inline jitify.hpp]: This will leak if an error occurs below. We could replace it with a std::vector<char>.

File: jitify.hpp [hunk 2]
```diff
@@ -1535,6 +1650,23 @@ static const char* jitsafe_header_math =
     // Note: Global namespace already includes CUDA math funcs
     "//using namespace __jitify_math_ns;\n";
 
+// TODO: incomplete
+static const char* jitsafe_header_mutex = R"(
```
Comment (benbarsdell) [inline jitify.hpp]: Is this actually needed in device code somewhere? (vs. just being an ignored header)

Comment (maddyscientist) [reply jitify.hpp]: This was needed for one of the cub includes.  Adding this silenced that.

**Pull request merged**